lib/rmg/RmgArea.cpp

```diff
@@ -0,0 +1,395 @@
+ * Full text of license available in license.txt file, in main folder
+ *
+ */
+
+#include "RmgArea.h"
+#include "CMapGenerator.h"
+
+namespace rmg
+{
+
+void toAbsolute(Tileset & tiles, const int3 & position)
+{
+	Tileset temp;
+	for(auto & tile : tiles)
+	{
+		temp.insert(tile + position);
+	}
+	tiles = std::move(temp);
+}
+
+void toRelative(Tileset & tiles, const int3 & position)
+{
+	toAbsolute(tiles, -position);
+}
+
+Area::Area(const Area & area): dTiles(area.dTiles), dTotalShiftCache(area.dTotalShiftCache)
+{
+}
+
+Area::Area(const Area && area): dTiles(std::move(area.dTiles)), dTotalShiftCache(std::move(area.dTotalShiftCache))
+{
+}
+
+
+Area & Area::operator=(const Area & area)
+{
+	clear();
+	dTiles = area.dTiles;
+	dTotalShiftCache = area.dTotalShiftCache;
+	return *this;
+}
+
+Area::Area(const Tileset & tiles): dTiles(tiles)
+{
+}
+
+Area::Area(const Tileset & relative, const int3 & position): dTiles(relative), dTotalShiftCache(position)
+{
+}
+
+void Area::invalidate()
+{
+	getTiles();
+	dTilesVectorCache.clear();
+	dBorderCache.clear();
+	dBorderOutsideCache.clear();
+}
+
+bool Area::connected() const
+{
+	std::list<int3> queue({*dTiles.begin()});
+	Tileset connected = dTiles; //use invalidated cache - ok
+	while(!queue.empty())
+	{
+		auto t = queue.front();
+		connected.erase(t);
+		queue.pop_front();
+		
+		for(auto & i : int3::getDirs())
+		{
+			if(connected.count(t + i))
+			{
+				queue.push_back(t + i);
+			}
+		}
+	}
+	
+	return connected.empty();
+}
+
+std::list<Area> connectedAreas(const Area & area)
+{
+	std::list<Area> result;
+	Tileset connected = area.getTiles();
+	while(!connected.empty())
+	{
+		result.emplace_back();
+		std::list<int3> queue({*connected.begin()});
+		std::set<int3> queueSet({*connected.begin()});
+		while(!queue.empty())
+		{
+			auto t = queue.front();
+			connected.erase(t);
+			result.back().add(t);
+			queue.pop_front();
+			
+			for(auto & i : int3::getDirs())
+			{
+				auto tile = t + i;
+				if(!queueSet.count(tile) && connected.count(tile) && !result.back().contains(tile))
+				{
+					queueSet.insert(tile);
+					queue.push_back(tile);
+				}
+			}
+		}
+	}
+	return result;
+}
+
+const Tileset & Area::getTiles() const
+{
+	if(dTotalShiftCache != int3())
+	{
+		toAbsolute(dTiles, dTotalShiftCache);
+		dTotalShiftCache = int3();
+	}
+	return dTiles;
+}
+
+const std::vector<int3> & Area::getTilesVector() const
+{
+	if(dTilesVectorCache.empty())
+	{
+		getTiles();
+		dTilesVectorCache.assign(dTiles.begin(), dTiles.end());
+	}
+	return dTilesVectorCache;
+}
+
+const Tileset & Area::getBorder() const
+{
+	if(!dBorderCache.empty())
+		return dBorderCache;
+	
+	//compute border cache
+	for(auto & t : dTiles)
+	{
+		for(auto & i : int3::getDirs())
+		{
+			if(!dTiles.count(t + i))
+				dBorderCache.insert(t + dTotalShiftCache);
+		}
+	}
+	
+	return dBorderCache;
+}
+
+const Tileset & Area::getBorderOutside() const
+{
+	if(!dBorderOutsideCache.empty())
```

Comment:
In some cases it might be faster if getBorderOutside calls getBorder if we use both methods for one tileset quite often

Reply:
Or even better we can calculate both vectors in one pass! Not sure which way is better. Depends on how often we use both border and border-outside for the same Area

Reply:
Computing them together could be interesting option, let me investigate if we could save something